opal/util/opal_gethostname.c

```diff
@@ -0,0 +1,178 @@
+            if( 0 == ret_val ) {
+                count = strlen( buf );
+                /*The result was not truncated */
+                if( count > 0 && count < length - 1 ) {
+                    break;
+                }
+                /*
+                 * The buffer is empty. In some gethostname
+                 * implementations this can be because the buffer was too 
+                 * small. But, if it is zero for some other reason, we
+                 * could get caught in the loop. So, break after
+                 * NUM_TRIES_FOR_NULL_HOSTNAME. 
+                 */
+                else if( 0 == count ) {
+                    if( num_tries > NUM_TRIES_FOR_NULL_HOSTNAME ) {
+                        ret_val = -1;
+                        temp_errno = EINVAL;
+                        break;
+                    }
+                    num_tries++;
+                }
+                /* The result *may* be truncated (can't know for sure, so we increase 
+                 * the length of the buffer (happens below) and try again). */
+                else if( count == length - 1 ) {
+			; /* no op */
+                }
+                /* Unidentified case, implies some bug in this code or some unknown
+                 * return string from gethostname() */
+                else {
+                    strcpy( buf, UNKNOWN_ERROR_STRING );
+                    break;
+                }
+            }
+            /*
+             * hostname was truncated and there was an error. Perhaps
+             * there is something in the buffer and perhaps not. 
+             * Set temp_errno to the most appropriate value;
+             * Double the length and try again. 
+             */
+            else if( EINVAL == errno || ENAMETOOLONG == errno ) {
+                temp_errno = ENAMETOOLONG;
+            }
+            /*
+             * Memory error (invalid address). Have we run out of memory?
+             */
+            else if( EFAULT == errno ) {
+                temp_errno = errno;
+                strcpy( buf, MEMORY_ERROR_STRING );
+                break;
+            }
+            /* Unidentified case, implies some unknown error returned from gethostname() */
+            else {
+                temp_errno = errno;
+                strcpy( buf, UNKNOWN_ERROR_STRING );
+                break;
+            }
+            length *= 2;
+            buf = realloc( buf, length );
+            if( NULL == buf ) {
+               errno = ENOMEM;
+               return MALLOC_ERROR_STRING;
+            }
+        }
+        /*
+         * Eventually, we would like this code to allow the full string
+         * to be copied in all (or all reasonable) cases (arbitrary length)
+         */
+        strncpy( opal_local_hostname, buf, OPAL_LOCAL_MAXHOSTNAMELEN - 1 );
+        opal_local_hostname[OPAL_LOCAL_MAXHOSTNAMELEN - 1] = '\0';
+        free( buf );
+        /* restore errno if there was a prior error to preserve error condition */
```

Comment:
Per my other comment, I think that since `gethostname()` may write a new value to `errno`, there is no expectation that a prior value of `errno` will be preserved after calling `gethostname()`.  So it's probably ok to not save a prior value of `errno`.